crypto/fipsmodule/ml_kem/ml_kem_test.cc

```diff
@@ -0,0 +1,1628 @@
+  0xf8, 0x8c, 0xb2, 0x5f, 0x89, 0xa3, 0x55, 0x5f, 0xae, 0xc6, 0x71, 0xa1,
+  0xdf, 0xc6, 0xf6, 0x1d, 0x60, 0xd0, 0x62, 0x22, 0x7d, 0x6a, 0x8f, 0xf6,
+  0x2b, 0x3c, 0x6d, 0x7b, 0xd6, 0x14, 0x0f, 0x66, 0x24, 0xc0, 0x84, 0xa6,
+  0x4d, 0xa7, 0x4c, 0x63, 0x32, 0x7e, 0x11, 0x77, 0x58, 0xaa, 0x33, 0x8a,
+  0x02, 0xe4, 0x43, 0x74, 0x10, 0xb8, 0xf9, 0xf2, 0x00, 0x88, 0xa1, 0x29,
+  0xc1, 0x68, 0x3d, 0xe7
+};
+
+struct MLKEMKeypairTestVector {
+  size_t public_len;
+  size_t secret_len;
+  const uint8_t *seed;
+  int (*keypair_deterministic)(uint8_t *public_key, size_t *public_len, uint8_t *secret_key, size_t *secret_len, const uint8_t *seed);
+  int (*keypair)(uint8_t *public_key, size_t *public_len, uint8_t *secret_key, size_t *secret_len);
+};
+
+
+static constexpr MLKEMKeypairTestVector keypairParameterSet[] = {
+  {
+    MLKEM512_PUBLIC_KEY_BYTES,
+    MLKEM512_SECRET_KEY_BYTES,
+    keyGen512Seed,
+    ml_kem_512_keypair_deterministic,
+    ml_kem_512_keypair
+  },
+  {
+    MLKEM768_PUBLIC_KEY_BYTES,
+    MLKEM768_SECRET_KEY_BYTES,
+    keyGen768Seed,
+    ml_kem_768_keypair_deterministic,
+    ml_kem_768_keypair
+  },
+  {
+    MLKEM1024_PUBLIC_KEY_BYTES,
+    MLKEM1024_SECRET_KEY_BYTES,
+    keyGen1024Seed,
+    ml_kem_1024_keypair_deterministic,
+    ml_kem_1024_keypair
+  }
+};
+
+class MLKEMKeypairLengthTest : public testing::TestWithParam<MLKEMKeypairTestVector> {
+public:
+  void TearDown() override {
+    delete [] encaps;
+    delete [] decaps;
+  }
+
+  uint8_t *encaps;
+  uint8_t *decaps;
+};
+
+INSTANTIATE_TEST_SUITE_P(All, MLKEMKeypairLengthTest, testing::ValuesIn(keypairParameterSet));
+
+TEST_P(MLKEMKeypairLengthTest, ExactLengthDeterministic) {
+  MLKEMKeypairTestVector params = GetParam();
+  size_t encaps_len = params.public_len;
+  encaps = new uint8_t[encaps_len];
+  size_t decaps_len = params.secret_len;
+  decaps = new uint8_t[decaps_len];
+  ASSERT_EQ(0, params.keypair_deterministic(encaps, &encaps_len, decaps, &decaps_len, params.seed));
+  ASSERT_EQ(params.public_len, encaps_len);
+  ASSERT_EQ(params.secret_len, decaps_len);
+}
+
+TEST_P(MLKEMKeypairLengthTest, ExactLength) {
+  MLKEMKeypairTestVector params = GetParam();
+  size_t encaps_len = params.public_len;
+  encaps = new uint8_t[encaps_len];
+  size_t decaps_len = params.secret_len;
+  decaps = new uint8_t[decaps_len];
+  ASSERT_EQ(0, params.keypair(encaps, &encaps_len, decaps, &decaps_len));
+  ASSERT_EQ(params.public_len, encaps_len);
+  ASSERT_EQ(params.secret_len, decaps_len);
+}
+
+TEST_P(MLKEMKeypairLengthTest, AllShortLengthDeterministic) {
+  MLKEMKeypairTestVector params = GetParam();
+  size_t encaps_len = params.public_len/2;
+  encaps = new uint8_t[encaps_len];
+  size_t decaps_len = params.secret_len/2;
+  decaps = new uint8_t[decaps_len];
+  ASSERT_NE(0, params.keypair_deterministic(encaps, &encaps_len, decaps, &decaps_len, params.seed));
+  ASSERT_EQ(params.public_len/2, encaps_len);
+  ASSERT_EQ(params.secret_len/2, decaps_len);
```

Comment:
Double checking: we're saying the low level keypair_determinisitic does not change output lengths if they're too small? Can we make the length a const pointer or just pass in a regular int?